### PR TITLE
Don't return logical slots for standby cluster

### PR DESCRIPTION
**Pull request by CyberDem0n**
Cluster.get_replication_slots() didn't take into account that there can not be logical replication slots in a standby cluster replicas. It was only skipping logical slots for the standby_leader, but replicas were expecting that they will have to copy them over.

Also on replicas in a standby cluster these logical slots were falsely added to the `_replication_slots` dict.

### CODE REVIEW

File: patroni/dcs/__init__.py [hunk 1]
```diff
@@ -1022,33 +1024,32 @@ def _merge_permanent_slots(slots: Dict[str, Dict[str, str]], permanent_slots: Di
...
+    def _get_permanent_slots(self, is_standby_cluster: bool, role: str, nofailover: bool) -> Dict[str, Any]:
         """Get configured permanent slot names.
 
         .. note::
-            Permanent logical replication slots are only considered if ``use_slots`` configuration is enabled. Also,
-            only considered if *role* is ``primary`` or if it is a promotable ``replica`` -- what excludes a
-            ``standby_leader`` or ``replica`` with ``nofailover`` tag enabled. That combination is used for failing
-            over logical replication slots, and the latter nodes are not eligible for such task.
+            Permanent replication slots are only considered if ``use_slots`` configuration is enabled.
+            Node that is not supposed to become a leader (nofailover) will not have permanent replication slots.
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1032]:
```suggestion
            A node that is not supposed to become a leader (*nofailover*) will not have permanent replication slots.
```

File: patroni/dcs/__init__.py [hunk 2]
```diff
@@ -1022,33 +1024,32 @@ def _merge_permanent_slots(slots: Dict[str, Dict[str, str]], permanent_slots: Di
...
 
-            Permanent physical slots are only considered if *role* is ``primary`` or ``standby_leader``, independently
-            if ``use_slots`` is enabled or not. That is done that way because even if Patroni itself is not using slots
-            to replicate among its members when ``use_slots`` is disabled, the user may still have configured Patroni to
-            keep permanent physical slots used out of Patroni.
+            In a standby cluster we support only physical replication slots.
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1034]:
```suggestion
            In a standby cluster we only support physical replication slots.
```

File: patroni/dcs/__init__.py [hunk 3]
```diff
@@ -1022,33 +1024,32 @@ def _merge_permanent_slots(slots: Dict[str, Dict[str, str]], permanent_slots: Di
...
 
+            For non-standby cluster returned dictionary always contains permanent logical replication slots in order
+            to show a warning if they are not supported by PostgreSQL before v11.
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1037]:
```suggestion
            The returned dictionary for a non-standby cluster always contains permanent logical replication slots in
            order to show a warning if they are not supported by PostgreSQL before v11.
```

File: patroni/postgresql/__init__.py [hunk 4]
```diff
@@ -413,7 +413,18 @@ def reset_cluster_info_state(self, cluster: Union[Cluster, None], nofailover: bo
...
-        if cluster and cluster.config and cluster.config.modify_version:
+
+        if global_config:
+            self._global_config = global_config
+
+        if not self._global_config:
+            return
+
+        if self._global_config.is_standby_cluster:
+            # standby cluster can't have logical replication slots and we don't need to enforce hot_standby_feedback
```
Comment (matthbakeredb) [inline patroni/postgresql/__init__.py:424]:
```suggestion
            # Standby cluster can't have logical replication slots, and we don't need to enforce hot_standby_feedback
```

File: patroni/postgresql/slots.py [hunk 5]
```diff
@@ -475,6 +475,11 @@ class instance. Slots that exist are also advanced if their ``confirmed_flush_ls
             elif cluster.slots and name in cluster.slots:  # We want to copy only slots with feedback in a DCS
                 create_slots.append(name)
 
+        # slots to be copied from the primary should be removed from the structure,
```
Comment (matthbakeredb) [inline patroni/postgresql/slots.py:478]:
```suggestion
        # Slots to be copied from the primary should be removed from the *slots* structure,
```

File: patroni/postgresql/slots.py [hunk 6]
```diff
@@ -497,12 +502,13 @@ def sync_replication_slots(self, cluster: Cluster, nofailover: bool,
...
-        if self._postgresql.major_version >= 90400 and cluster.config:
+        if self._postgresql.major_version >= 90400 and self._postgresql.global_config and cluster.config:
             try:
                 self.load_replication_slots()
 
-                slots = cluster.get_replication_slots(self._postgresql.name, self._postgresql.role,
-                                                      nofailover, self._postgresql.major_version, True)
+                slots = cluster.get_replication_slots(
+                    self._postgresql.name, self._postgresql.role, nofailover, self._postgresql.major_version,
+                    is_standby_cluster=self._postgresql.global_config.is_standby_cluster, show_error=True)
```
Comment (matthbakeredb) [inline patroni/postgresql/slots.py:511]:
```suggestion
                    self._postgresql.name,
                    self._postgresql.role,
                    nofailover,
                    self._postgresql.major_version,
                    is_standby_cluster=self._postgresql.global_config.is_standby_cluster,
                    show_error=True,
                )
```

**Review by matthbakeredb — CHANGES_REQUESTED** [review]
Looks good, small formatting, comment suggestions

**Review by barthisrael — APPROVED** [review]
Looks good to me too.

File: patroni/dcs/__init__.py [hunk 7]
```diff
@@ -932,8 +932,8 @@ def use_slots(self) -> bool:
         """``True`` if cluster is configured to use replication slots."""
         return bool(self.config and (self.config.data.get('postgresql') or {}).get('use_slots', True))
 
-    def get_replication_slots(self, my_name: str, role: str, nofailover: bool,
-                              major_version: int, show_error: bool = False) -> Dict[str, Dict[str, Any]]:
+    def get_replication_slots(self, my_name: str, role: str, nofailover: bool, major_version: int, *,
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:935]: I changed the order of the arguments with default values and it took me some time to figure out why tests are failing, while linting didn't found any problem.

[the remark this answers is not in the document]
Comment (barthisrael) [reply patroni/dcs/__init__.py:935]: Ah, cool then, thanks!

File: patroni/postgresql/slots.py (hunk 6 above)
Comment (matthbakeredb) [reply patroni/postgresql/slots.py:511]: The goal is not to keep the code shorter using less space but for me to easily discern between args. It's easier on my eye to scan down a list than to read horizontally with mixed breaks in dividing spaces.

btw I thought this looked a little messy and ran black on it and this was the result.

File: patroni/dcs/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:935]: In $previous_job we had a local rule that all methods should be explicitly called with full keyword args to ensure that if the signature changed in the called method the calling code wouldn't start sending an input to the wrong place.

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**